subsys/bluetooth/controller/hci/hci.c

```diff
@@ -4180,6 +4188,207 @@ static void le_per_adv_sync_established(struct pdu_data *pdu_data,
+			ptr++;
+
+			BT_DBG("    Tx pwr= %d dB", tx_pwr);
+		}
+
+		hdr_len = ptr - (uint8_t *)p;
+		if (hdr_len <= (offsetof(struct pdu_adv_com_ext_adv,
+					 ext_hdr_adi_adv_data) +
+				sizeof(struct pdu_adv_hdr))) {
+			hdr_len = offsetof(struct pdu_adv_com_ext_adv,
+					   ext_hdr_adi_adv_data);
+			ptr = (uint8_t *)h;
+		}
+
+		if (hdr_len > (p->ext_hdr_len +
+			       offsetof(struct pdu_adv_com_ext_adv,
+					ext_hdr_adi_adv_data))) {
+			BT_WARN("    Header length %u/%u, INVALID.", hdr_len,
+				p->ext_hdr_len);
+		} else {
+			uint8_t acad_len = p->ext_hdr_len +
+					   offsetof(struct pdu_adv_com_ext_adv,
+						    ext_hdr_adi_adv_data) -
+					   hdr_len;
+
+			if (acad_len) {
+				ptr += acad_len;
+				hdr_len += acad_len;
+
+				BT_DBG("ACAD: <todo>");
+			}
+		}
+
+no_ext_hdr:
+		if (hdr_len < adv->len) {
+			data_len_curr = adv->len - hdr_len;
+			data_curr = ptr;
+
+			BT_DBG("    AD Data (%u): <todo>", data_len);
+		}
+
+		if (node_rx_curr == node_rx) {
+			data_len = data_len_curr;
+			total_data_len = data_len;
+			data = data_curr;
+		} else {
+			/* TODO: Validate current value with previous ??
+			 */
+
+			if (!data) {
+				data_len = data_len_curr;
+				total_data_len = data_len;
+				data = data_curr;
+			} else {
+				total_data_len += data_len_curr;
+
+				/* TODO: construct new HCI event for this
+				 * fragment.
+				 */
+			}
+		}
+
+		if (!node_rx_next) {
+			if (sec_phy_curr) {
+				data_status =
+				  BT_HCI_LE_ADV_EVT_TYPE_DATA_STATUS_INCOMPLETE;
+			}
+
+			break;
+		}
+
+		node_rx_curr = node_rx_next;
+		node_rx_next = node_rx_curr->hdr.rx_ftr.extra;
+		adv = (void *)node_rx_curr->pdu;
+	} while (1);
+
+	/* FIXME: move most of below into above loop to dispatch fragments of
+	 * data in HCI event.
+	 */
+
+	/* If data complete */
+	if (!data_status) {
+		uint8_t data_max_len;
+
+		data_max_len = CONFIG_BT_DISCARDABLE_BUF_SIZE -
+			       BT_HCI_ACL_HDR_SIZE - sizeof(*sep);
+
+		/* if data cannot fit the event, mark it as incomplete */
+		if (data_len > data_max_len) {
+			data_len = data_max_len;
+			data_status =
+				BT_HCI_LE_ADV_EVT_TYPE_DATA_STATUS_PARTIAL;
+		}
+	} else {
+		/* Data incomplete and no more to come */
+		if ((tx_pwr == BT_HCI_LE_ADV_TX_POWER_NO_PREF) && !data) {
+			goto le_per_adv_report_invalid;
+		}
+	}
+
+	/* Start constructing the event */
+	sep = meta_evt(buf, BT_HCI_EVT_LE_PER_ADVERTISING_REPORT,
+		       sizeof(*sep) + data_len);
+
+	sep->handle = sys_cpu_to_le16(node_rx->hdr.handle);
+	sep->tx_power = tx_pwr;
+	sep->rssi = rssi;
+	sep->cte_type = 0U; /* TODO */
+	sep->data_status = data_status;
+	sep->length = data_len;
+	memcpy(&sep->data[0], data, data_len);
+
+le_per_adv_report_invalid:
+	/* Free the node_rx list */
+	node_rx_next = node_rx->hdr.rx_ftr.extra;
+	while (node_rx_next) {
+		node_rx_curr = node_rx_next;
+		node_rx_next = node_rx_curr->hdr.rx_ftr.extra;
+
+		node_rx_curr->hdr.next = NULL;
+		ll_rx_mem_release((void **)&node_rx_curr);
+	}
```

Comment:
Please extract this into a separate function that can be shared with `le_ext_adv_report` and avoid the goto.

Reply:
done.